// cdk-monitoring-constructs/lib/monitoring/aws-docdb/DocumentDbMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DocumentDbMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
const DocumentDbNamespace = "AWS/DocDB";
class DocumentDbMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.clusterIdentifier = props.cluster.clusterIdentifier;
        this.dimensionsMap = { DBClusterIdentifier: this.clusterIdentifier };
    }
    metricAverageCpuUsageInPercent() {
        return this.metric("CPUUtilization", common_1.MetricStatistic.AVERAGE, "CPU Usage");
    }
    metricMaxConnectionCount() {
        return this.metric("DatabaseConnectionsMax", common_1.MetricStatistic.MAX, "Connections");
    }
    metricMaxCursorCount() {
        return this.metric("DatabaseCursorsMax", common_1.MetricStatistic.MAX, "Cursors");
    }
    metricMaxTransactionOpenCount() {
        return this.metric("TransactionsOpenMax", common_1.MetricStatistic.MAX, "Transactions");
    }
    metricOperationsThrottledDueLowMemoryCount() {
        return this.metric("LowMemNumOperationsThrottled", common_1.MetricStatistic.SUM, "Operations throttled (low mem)");
    }
    metricReadLatencyInMillis(latencyType) {
        return this.metric("ReadLatency", common_1.getLatencyTypeStatistic(latencyType), `Read ${common_1.getLatencyTypeLabel(latencyType)}`);
    }
    metricWriteLatencyInMillis(latencyType) {
        return this.metric("WriteLatency", common_1.getLatencyTypeStatistic(latencyType), `Write ${common_1.getLatencyTypeLabel(latencyType)}`);
    }
    metric(metricName, statistic, label) {
        return this.metricFactory.createMetric(metricName, statistic, label, this.dimensionsMap, undefined, DocumentDbNamespace, undefined, this.region, this.account);
    }
}
exports.DocumentDbMetricFactory = DocumentDbMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
DocumentDbMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.DocumentDbMetricFactory", version: "8.1.0" };
