// cdk-monitoring-constructs/lib/monitoring/aws-lambda/LambdaFunctionEnhancedMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LambdaFunctionEnhancedMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
const LambdaInsightsNamespace = "LambdaInsights";
class LambdaFunctionEnhancedMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.lambdaFunction = props.lambdaFunction;
    }
    enhancedMetricMaxCpuTotalTime() {
        return this.enhancedMetric("cpu_total_time", common_1.MetricStatistic.MAX, "CPUTotalTime.Max");
    }
    enhancedMetricP90CpuTotalTime() {
        return this.enhancedMetric("cpu_total_time", common_1.MetricStatistic.P90, "CPUTotalTime.P90");
    }
    enhancedMetricAvgCpuTotalTime() {
        return this.enhancedMetric("cpu_total_time", common_1.MetricStatistic.AVERAGE, "CPUTotalTime.Avg");
    }
    enhancedMetricMaxMemoryUtilization() {
        return this.enhancedMetric("memory_utilization", common_1.MetricStatistic.MAX, "MemoryUtilization.Max");
    }
    enhancedMetricP90MemoryUtilization() {
        return this.enhancedMetric("memory_utilization", common_1.MetricStatistic.P90, "MemoryUtilization.P90");
    }
    enhancedMetricAvgMemoryUtilization() {
        return this.enhancedMetric("memory_utilization", common_1.MetricStatistic.AVERAGE, "MemoryUtilization.Avg");
    }
    enhancedMetricFunctionCost() {
        return this.metricFactory.createMetricMath("memory_utilization * duration", {
            memory_utilization: this.enhancedMetricMaxMemoryUtilization(),
            duration: this.enhancedMetricFunctionDuration(),
        }, "Function Cost (avg: ${AVG}, max: ${MAX})");
    }
    enhancedMetricFunctionDuration() {
        return this.metricFactory.adaptMetric(this.lambdaFunction.metricDuration({
            statistic: common_1.MetricStatistic.SUM,
            region: this.region,
            account: this.account,
        }));
    }
    enhancedMetric(metricName, statistic, label, color) {
        const [functionName, functionVersion] = this.lambdaFunction.functionName.split(":");
        return this.metricFactory.createMetric(metricName, statistic, label, {
            function_name: functionName,
            version: functionVersion,
        }, color, LambdaInsightsNamespace, undefined, this.region, this.account);
    }
}
exports.LambdaFunctionEnhancedMetricFactory = LambdaFunctionEnhancedMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
LambdaFunctionEnhancedMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.LambdaFunctionEnhancedMetricFactory", version: "8.1.0" };
