// cdk-monitoring-constructs/lib/monitoring/fluentbit/FluentBitMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.FluentBitMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_logs_1 = require("aws-cdk-lib/aws-logs");
const FluentBitConstants_1 = require("./FluentBitConstants");
const common_1 = require("../../common");
class FluentBitMetricFactory extends common_1.BaseMetricFactory {
    constructor(scope, props) {
        super(scope.createMetricFactory(), props);
        this.scope = scope;
        this.namespace =
            props.namespace ??
                this.metricFactory.getNamespaceWithFallback(props.namespace);
    }
    filterMetrics(logGroup) {
        return Object.values(FluentBitConstants_1.FluentBitFilterMetricTag).map((metricName) => {
            return this.pluginMetric(logGroup, metricName);
        });
    }
    outputMetrics(logGroup) {
        return Object.values(FluentBitConstants_1.FluentBitOutputMetricTag).map((metricName) => {
            return this.pluginMetric(logGroup, metricName);
        });
    }
    inputMetrics(logGroup) {
        return Object.values(FluentBitConstants_1.FluentBitInputMetricTag).map((metricName) => {
            return this.pluginMetric(logGroup, metricName);
        });
    }
    pluginMetric(logGroup, metricName) {
        const metricFilter = new aws_logs_1.MetricFilter(this.scope, `FluentBit-${metricName}-${logGroup}-MetricFilter`, {
            logGroup: logGroup,
            filterPattern: aws_logs_1.FilterPattern.literal(`{ $.metric = "${metricName}" }`),
            metricNamespace: this.namespace,
            metricName,
            metricValue: "$.value",
        });
        return metricFilter.metric({
            statistic: common_1.MetricStatistic.MAX,
            region: this.region,
            account: this.account,
        });
    }
    storageMetrics(logGroup) {
        return Object.values(FluentBitConstants_1.FluentBitStorageMetricTag).map((metricName) => {
            const valueString = `$.storage_layer.chunks.${metricName}`;
            const metricFilter = new aws_logs_1.MetricFilter(this.scope, `FluentBit-${metricName}-${logGroup}-MetricFilter`, {
                logGroup: logGroup,
                filterPattern: aws_logs_1.FilterPattern.literal(`{ ${valueString} = * }`),
                metricNamespace: this.namespace,
                metricName,
                metricValue: `${valueString}`,
            });
            return metricFilter.metric({
                statistic: common_1.MetricStatistic.MAX,
                region: this.region,
                account: this.account,
            });
        });
    }
    metricsWithoutWidgets(logGroup) {
        Object.values(FluentBitConstants_1.FluentBitMetricsWithoutWidget).forEach((metricName) => this.pluginMetric(logGroup, metricName));
    }
}
exports.FluentBitMetricFactory = FluentBitMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
FluentBitMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.FluentBitMetricFactory", version: "8.1.0" };
