// cdk-monitoring-constructs/lib/monitoring/aws-codebuild/CodeBuildProjectMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeBuildProjectMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
class CodeBuildProjectMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.project = props.project;
        this.dimensionsMap = {
            ProjectName: props.project.projectName,
        };
    }
    metricBuildCount() {
        return this.metricFactory.adaptMetric(this.project.metricBuilds({
            region: this.region,
            account: this.account,
        }));
    }
    metricSucceededBuildCount() {
        return this.metricFactory.adaptMetric(this.project.metricSucceededBuilds({
            region: this.region,
            account: this.account,
        }));
    }
    metricFailedBuildCount() {
        return this.metricFactory.adaptMetric(this.project.metricFailedBuilds({
            region: this.region,
            account: this.account,
        }));
    }
    metricFailedBuildRate() {
        return this.metricFailedBuildCount().with({
            statistic: common_1.MetricStatistic.AVERAGE,
        });
    }
    metricDurationP99InSeconds() {
        return this.metricFactory.adaptMetric(this.project.metricDuration({
            label: "P99",
            statistic: common_1.MetricStatistic.P99,
            region: this.region,
            account: this.account,
        }));
    }
    metricDurationP90InSeconds() {
        return this.metricFactory.adaptMetric(this.project.metricDuration({
            label: "P90",
            statistic: common_1.MetricStatistic.P90,
            region: this.region,
            account: this.account,
        }));
    }
    metricDurationP50InSeconds() {
        return this.metricFactory.adaptMetric(this.project.metricDuration({
            label: "P50",
            statistic: common_1.MetricStatistic.P50,
            region: this.region,
            account: this.account,
        }));
    }
}
exports.CodeBuildProjectMetricFactory = CodeBuildProjectMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
CodeBuildProjectMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.CodeBuildProjectMetricFactory", version: "8.1.0" };
