// cdk-monitoring-constructs/lib/common/alarm/AlarmNamingStrategy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AlarmNamingStrategy = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const IAlarmDedupeStringProcessor_1 = require("./IAlarmDedupeStringProcessor");
const AlarmNamePartSeparator = "-";
const AlarmLabelPartSeparator = " ";
class AlarmNamingStrategy {
    constructor(globalPrefix, localPrefix, dedupeStringStrategy) {
        this.globalPrefix = globalPrefix;
        this.localPrefix = localPrefix;
        this.dedupeStringStrategy =
            dedupeStringStrategy ?? new IAlarmDedupeStringProcessor_1.DoNotModifyDedupeString();
    }
    /**
     * Alarm name is resolved like this:
     * - If "alarmNameOverride" is defined for an alarm, it will be used as alarm name.
     * - Otherwise, the alarm name will be generated by joining: global prefix, local prefix, alarm name suffix, disambiguator.
     *
     * @param props properties
     */
    getName(props) {
        if (props.alarmNameOverride) {
            return props.alarmNameOverride;
        }
        const parts = [
            this.globalPrefix,
            this.localPrefix,
            props.alarmNameSuffix,
        ];
        if (props.disambiguator) {
            parts.push(props.disambiguator);
        }
        return this.joinDistinct(parts, AlarmNamePartSeparator);
    }
    getWidgetLabel(props) {
        // not using global prefix to make the label shorter
        const parts = [this.localPrefix, props.alarmNameSuffix];
        if (props.disambiguator) {
            parts.push(props.disambiguator);
        }
        return this.joinDistinct(parts, AlarmLabelPartSeparator);
    }
    /**
     * Dedupe string resolved like this:
     * - If "dedupeStringOverride" is defined for an alarm, it will be used as a dedupe string.
     * - If "alarmDedupeStringSuffix" from the alarm factory is defined, "GlobalPrefix-LocalPrefix-AlarmDedupeStringSuffix" will be used as a dedupe string.
     * - Otherwise, the alarm dedupe string will not be set.
     * If a dedupe string strategy is set, it will be used to process the final string.
     *
     * @param props properties
     */
    getDedupeString(props) {
        if (props.dedupeStringOverride) {
            return this.dedupeStringStrategy.processDedupeStringOverride(props.dedupeStringOverride);
        }
        if (props.alarmDedupeStringSuffix) {
            // not using disambiguator here as we are still alarming on the same thing
            const parts = [
                this.globalPrefix,
                this.localPrefix,
                props.alarmDedupeStringSuffix,
            ];
            const dedupeString = this.joinDistinct(parts, AlarmNamePartSeparator);
            return this.dedupeStringStrategy.processDedupeString(dedupeString);
        }
        return undefined;
    }
    joinDistinct(parts, separator) {
        return parts.filter((x, i, a) => a.indexOf(x) == i).join(separator);
    }
}
exports.AlarmNamingStrategy = AlarmNamingStrategy;
_a = JSII_RTTI_SYMBOL_1;
AlarmNamingStrategy[_a] = { fqn: "cdk-monitoring-constructs.AlarmNamingStrategy", version: "8.1.0" };
