// cdk-monitoring-constructs/lib/common/widget/axis.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CurrencyAxisUsdFromZero = exports.MegabyteMillisecondAxisFromZero = exports.SizeAxisGigaBytesFromZero = exports.SizeAxisMegaBytesFromZero = exports.SizeAxisKiloBytesFromZero = exports.SizeAxisBytesFromZero = exports.RateAxisFromZeroToOne = exports.RateAxisFromZero = exports.BooleanAxisFromZeroToOne = exports.CountAxisFromOne = exports.CountAxisFromZero = exports.TimeAxisMinutesFromZero = exports.TimeAxisSecondsFromZero = exports.TimeAxisMillisFromZero = exports.PercentageAxisUpToHundred = exports.PercentageAxisFromZeroToHundred = exports.PercentageAxisFromZero = void 0;
/**
 * Y-Axis showing percentage, ranged from 0.
 */
exports.PercentageAxisFromZero = {
    min: 0,
    label: "%",
    showUnits: false,
};
/**
 * Y-Axis showing percentage, ranged from 0 to 100.
 */
exports.PercentageAxisFromZeroToHundred = {
    min: 0,
    max: 100,
    label: "%",
    showUnits: false,
};
/**
 * Y-Axis showing percentage, ranged up to 100.
 */
exports.PercentageAxisUpToHundred = {
    max: 100,
    label: "%",
    showUnits: false,
};
/**
 * Y-Axis showing time in milliseconds, ranged from 0.
 */
exports.TimeAxisMillisFromZero = {
    min: 0,
    label: "ms",
    showUnits: false,
};
/**
 * Y-Axis showing time in seconds, ranged from 0.
 */
exports.TimeAxisSecondsFromZero = {
    min: 0,
    label: "sec",
    showUnits: false,
};
/**
 * Y-Axis showing time in minutes, ranged from 0.
 */
exports.TimeAxisMinutesFromZero = {
    min: 0,
    label: "min",
    showUnits: false,
};
/**
 * Y-Axis showing count, ranged from 0.
 */
exports.CountAxisFromZero = {
    min: 0,
    label: "Count",
    showUnits: false,
};
/**
 * Y-Axis showing count, ranged from 1.
 */
exports.CountAxisFromOne = {
    min: 1,
    label: "Count",
    showUnits: false,
};
/**
 * Y-Axis showing boolean, ranged from 0 to 1, where 1 means TRUE.
 */
exports.BooleanAxisFromZeroToOne = {
    min: 0,
    max: 1,
    label: "1 = True",
    showUnits: false,
};
/**
 * Y-Axis showing rate (relative count), ranged from 0.
 */
exports.RateAxisFromZero = {
    min: 0,
    label: "Rate",
    showUnits: false,
};
/**
 * Y-Axis showing rate (relative count), ranged from 0 to 1.
 */
exports.RateAxisFromZeroToOne = {
    min: 0,
    max: 1,
    label: "Rate",
    showUnits: false,
};
/**
 * Y-Axis showing size in bytes, ranged from 0.
 */
exports.SizeAxisBytesFromZero = {
    min: 0,
    label: "bytes",
    showUnits: false,
};
/**
 * Y-Axis showing size in kilobytes, ranged from 0.
 */
exports.SizeAxisKiloBytesFromZero = {
    min: 0,
    label: "kB",
    showUnits: false,
};
/**
 * Y-Axis showing size in megabytes, ranged from 0.
 */
exports.SizeAxisMegaBytesFromZero = {
    min: 0,
    label: "MB",
    showUnits: false,
};
/**
 * Y-Axis showing size in gigabytes, ranged from 0.
 */
exports.SizeAxisGigaBytesFromZero = {
    min: 0,
    label: "GB",
    showUnits: false,
};
/**
 * Y-Axis showing cost in Megabyte Millisecond, ranged from 0.
 */
exports.MegabyteMillisecondAxisFromZero = {
    min: 0,
    label: "MB*ms",
    showUnits: false,
};
/**
 * Y-Axis showing cost in USD, ranged from 0.
 */
exports.CurrencyAxisUsdFromZero = {
    min: 0,
    label: "USD",
    showUnits: false,
};
