// cdk-monitoring-constructs/lib/common/widget/size.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.recommendedWidgetWidth = exports.DefaultLogWidgetHeight = exports.DefaultAlarmWidgetHeight = exports.DefaultAlarmWidgetWidth = exports.DefaultSummaryWidgetHeight = exports.DefaultTwoLinerGraphWidgetHalfHeight = exports.DefaultTwoLinerGraphWidgetHeight = exports.DefaultGraphWidgetHeight = exports.ThreeQuartersWidth = exports.TwoThirdsWidth = exports.SixthWidth = exports.HalfQuarterWidth = exports.QuarterWidth = exports.ThirdWidth = exports.HalfWidth = exports.FullWidth = void 0;
exports.FullWidth = 24;
exports.HalfWidth = exports.FullWidth / 2;
exports.ThirdWidth = exports.FullWidth / 3;
exports.QuarterWidth = exports.FullWidth / 4;
exports.HalfQuarterWidth = exports.QuarterWidth / 2;
exports.SixthWidth = exports.FullWidth / 6;
exports.TwoThirdsWidth = 2 * exports.ThirdWidth;
exports.ThreeQuartersWidth = 3 * exports.QuarterWidth;
// Widget Heights
exports.DefaultGraphWidgetHeight = 5;
exports.DefaultTwoLinerGraphWidgetHeight = 6;
exports.DefaultTwoLinerGraphWidgetHalfHeight = 3;
exports.DefaultSummaryWidgetHeight = 6;
exports.DefaultAlarmWidgetWidth = 6;
exports.DefaultAlarmWidgetHeight = 4;
exports.DefaultLogWidgetHeight = 7;
/**
 * Suggests the best widget width, given the total number of widgets.
 * The main point is to make widgets as wide as possible, while saving vertical space and minimizing number of gaps.
 *
 * @param numTotalWidgets total number of widgets to be placed
 */
function recommendedWidgetWidth(numTotalWidgets) {
    function numRowsTaken(numItems, itemSize) {
        return Math.ceil((numItems * itemSize) / exports.FullWidth);
    }
    const numItemsFixed = Math.max(1, numTotalWidgets);
    const widths = [exports.QuarterWidth, exports.ThirdWidth, exports.HalfWidth, exports.FullWidth];
    let i = 0;
    while (i < widths.length - 1 &&
        numRowsTaken(numItemsFixed, widths[i + 1]) ===
            numRowsTaken(numItemsFixed, widths[i])) {
        i++;
    }
    return widths[i];
}
exports.recommendedWidgetWidth = recommendedWidgetWidth;
