// cdk-monitoring-constructs/lib/common/widget/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createGraphWidget = exports.GraphWidgetType = void 0;
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
var GraphWidgetType;
(function (GraphWidgetType) {
    GraphWidgetType["BAR"] = "Bar";
    GraphWidgetType["LINE"] = "Line";
    GraphWidgetType["PIE"] = "Pie";
    GraphWidgetType["SINGLE_VALUE"] = "SingleValue";
    GraphWidgetType["STACKED_AREA"] = "StackedArea";
})(GraphWidgetType = exports.GraphWidgetType || (exports.GraphWidgetType = {}));
/**
 * Creates a graph widget of the desired type.
 *
 * @param type graph type (e.g. Pie or Bar)
 * @param props graph widget properties
 */
function createGraphWidget(type, props) {
    switch (type) {
        case GraphWidgetType.BAR:
            return new aws_cloudwatch_1.GraphWidget({
                ...props,
                view: aws_cloudwatch_1.GraphWidgetView.BAR,
            });
        case GraphWidgetType.LINE:
            return new aws_cloudwatch_1.GraphWidget(props);
        case GraphWidgetType.PIE:
            return new aws_cloudwatch_1.GraphWidget({
                ...props,
                view: aws_cloudwatch_1.GraphWidgetView.PIE,
            });
        case GraphWidgetType.SINGLE_VALUE:
            return new aws_cloudwatch_1.SingleValueWidget({
                ...props,
                metrics: [...(props.left ?? []), ...(props.right ?? [])],
            });
        case GraphWidgetType.STACKED_AREA:
            return new aws_cloudwatch_1.GraphWidget({
                ...props,
                stacked: true,
            });
        default:
            throw new Error(`Unsupported graph type: ${type}`);
    }
}
exports.createGraphWidget = createGraphWidget;
