// @cdklabs/generative-ai-cdk-constructs/lib/patterns/gen-ai/aws-langchain-common-layer/index.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.LangchainCommonLayer = exports.LangchainCommonDepsLayer = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const path = require("path");
const lambda = require("aws-cdk-lib/aws-lambda");
const constructs_1 = require("constructs");
const python_lambda_layer_helper_1 = require("../../../common/helpers/python-lambda-layer-helper");
/**
   * @summary The LangchainCommonDepsLayer class.
   */
class LangchainCommonDepsLayer extends constructs_1.Construct {
    /**
       * @summary This construct creates a lambda layer loaded with relevant libraries to run genai applications. Libraries include boto3, botocore, requests, requests-aws4auth, langchain, opensearch-py and openai.
       * @param {cdk.App} scope - represents the scope for all the resources.
       * @param {string} id - this is a a scope-unique id.
       * @param {LangchainLayerProps} props - user provided props for the construct.
       * @since 0.0.0
       * @access public
       */
    constructor(scope, id, props) {
        super(scope, id);
        const layer = new python_lambda_layer_helper_1.Layer(this, 'Langchain Layer', {
            path: path.join(__dirname, '../../../../layers/langchain-common-deps'),
            description: 'Dependencies to build gen ai applications with the langchain client',
            ...props,
        });
        this.layer = layer.layer;
    }
}
exports.LangchainCommonDepsLayer = LangchainCommonDepsLayer;
_a = JSII_RTTI_SYMBOL_1;
LangchainCommonDepsLayer[_a] = { fqn: "@cdklabs/generative-ai-cdk-constructs.LangchainCommonDepsLayer", version: "0.1.122" };
/**
   * @summary LangchainCommonLayer allows developers to instantiate a llm client adapter on bedrock, sagemaker or openai following best practise.
   *
   * @example
   * import boto3
   * from genai_core.adapters.registry import registry
   *
   * adapter = registry.get_adapter(f"{provider}.{model_id}")
   * bedrock_client = boto3.client('bedrock-runtime')
   */
class LangchainCommonLayer extends constructs_1.Construct {
    /**
       * @summary This construct allows developers to instantiate a llm client adapter on bedrock, sagemaker or openai following best practise.
       * @param {cdk.App} scope - represents the scope for all the resources.
       * @param {string} id - this is a a scope-unique id.
       * @param {AdapterProps} props - user provided props for the construct.
       * @since 0.0.0
       * @access public
       */
    constructor(scope, id, props) {
        super(scope, id);
        const layer = new lambda.LayerVersion(this, 'Model Adapter Layer', {
            code: lambda.Code.fromAsset(path.join(__dirname, '../../../../layers/langchain-common-layer')),
            description: 'Utilities to instantiate a llm client adapter. Adapters include bedrock, sagemaker, and openai',
            ...props,
        });
        this.layer = layer;
    }
}
exports.LangchainCommonLayer = LangchainCommonLayer;
_b = JSII_RTTI_SYMBOL_1;
LangchainCommonLayer[_b] = { fqn: "@cdklabs/generative-ai-cdk-constructs.LangchainCommonLayer", version: "0.1.122" };
