// cdk-monitoring-constructs/lib/common/alarm/action/MultipleAlarmActionStrategy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultipleAlarmActionStrategy = exports.isMultipleAlarmActionStrategy = exports.multipleActions = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
function multipleActions(...actions) {
    return new MultipleAlarmActionStrategy(actions);
}
exports.multipleActions = multipleActions;
function isMultipleAlarmActionStrategy(obj) {
    return !!(obj && obj instanceof MultipleAlarmActionStrategy);
}
exports.isMultipleAlarmActionStrategy = isMultipleAlarmActionStrategy;
/**
 * Alarm action strategy that combines multiple actions in the same order as they were given.
 */
class MultipleAlarmActionStrategy {
    constructor(actions) {
        this.actions = actions;
    }
    addAlarmActions(props) {
        this.actions.forEach((action) => action.addAlarmActions(props));
    }
    /**
     * Returns list of alarm actions where any nested instances of MultipleAlarmActionStrategy
     * are flattened.
     *
     * @returns flattened list of alarm actions.
     */
    flattenedAlarmActions() {
        return this._flattenedAlarmActions(...this.actions);
    }
    _flattenedAlarmActions(...actions) {
        return actions.flatMap((action) => {
            if (isMultipleAlarmActionStrategy(action)) {
                return this._flattenedAlarmActions(action);
            }
            return [action];
        });
    }
}
exports.MultipleAlarmActionStrategy = MultipleAlarmActionStrategy;
_a = JSII_RTTI_SYMBOL_1;
MultipleAlarmActionStrategy[_a] = { fqn: "cdk-monitoring-constructs.MultipleAlarmActionStrategy", version: "8.1.0" };
