// cdk-monitoring-constructs/lib/common/alarm/action/OpsItemAlarmActionStrategy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.OpsItemAlarmActionStrategy = exports.createOpsItem = exports.createLowSeverityOpsItem = exports.createMediumSeverityOpsItem = exports.createHighSeverityOpsItem = exports.createCriticalSeverityOpsItem = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_actions_1 = require("aws-cdk-lib/aws-cloudwatch-actions");
/**
 * Creates an AWS OpsCenter OpsItem with critical severity.
 *
 * @param category optional category (no category by default)
 */
function createCriticalSeverityOpsItem(category) {
    return createOpsItem(aws_cloudwatch_actions_1.OpsItemSeverity.CRITICAL, category);
}
exports.createCriticalSeverityOpsItem = createCriticalSeverityOpsItem;
/**
 * Creates an AWS OpsCenter OpsItem with high severity.
 *
 * @param category optional category (no category by default)
 */
function createHighSeverityOpsItem(category) {
    return createOpsItem(aws_cloudwatch_actions_1.OpsItemSeverity.HIGH, category);
}
exports.createHighSeverityOpsItem = createHighSeverityOpsItem;
/**
 * Creates an AWS OpsCenter OpsItem with medium severity.
 *
 * @param category optional category (no category by default)
 */
function createMediumSeverityOpsItem(category) {
    return createOpsItem(aws_cloudwatch_actions_1.OpsItemSeverity.MEDIUM, category);
}
exports.createMediumSeverityOpsItem = createMediumSeverityOpsItem;
/**
 * Creates an AWS OpsCenter OpsItem with low severity.
 *
 * @param category optional category (no category by default)
 */
function createLowSeverityOpsItem(category) {
    return createOpsItem(aws_cloudwatch_actions_1.OpsItemSeverity.LOW, category);
}
exports.createLowSeverityOpsItem = createLowSeverityOpsItem;
/**
 * Creates an AWS OpsCenter OpsItem.
 *
 * @param severity desired item severity
 * @param category optional category (no category by default)
 */
function createOpsItem(severity, category) {
    return new OpsItemAlarmActionStrategy(severity, category);
}
exports.createOpsItem = createOpsItem;
/**
 * Alarm action strategy that creates an AWS OpsCenter OpsItem.
 */
class OpsItemAlarmActionStrategy {
    constructor(severity, category) {
        this.severity = severity;
        this.category = category;
    }
    addAlarmActions(props) {
        props.alarm.addAlarmAction(new aws_cloudwatch_actions_1.SsmAction(this.severity, this.category));
    }
}
exports.OpsItemAlarmActionStrategy = OpsItemAlarmActionStrategy;
_a = JSII_RTTI_SYMBOL_1;
OpsItemAlarmActionStrategy[_a] = { fqn: "cdk-monitoring-constructs.OpsItemAlarmActionStrategy", version: "8.1.0" };
