// cdk-monitoring-constructs/lib/common/alarm/action/SnsAlarmActionStrategy.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnsAlarmActionStrategy = exports.notifySns = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_actions_1 = require("aws-cdk-lib/aws-cloudwatch-actions");
function notifySns(onAlarmTopic, onOkTopic, onInsufficientDataTopic) {
    return new SnsAlarmActionStrategy({
        onAlarmTopic,
        onOkTopic,
        onInsufficientDataTopic,
    });
}
exports.notifySns = notifySns;
/**
 * Alarm action strategy that sends a notification to the specified SNS topic.
 */
class SnsAlarmActionStrategy {
    constructor(props) {
        this.onAlarmTopic = props.onAlarmTopic;
        this.onOkTopic = props.onOkTopic;
        this.onInsufficientDataTopic = props.onInsufficientDataTopic;
    }
    addAlarmActions(props) {
        props.alarm.addAlarmAction(new aws_cloudwatch_actions_1.SnsAction(this.onAlarmTopic));
        if (this.onOkTopic) {
            props.alarm.addOkAction(new aws_cloudwatch_actions_1.SnsAction(this.onOkTopic));
        }
        if (this.onInsufficientDataTopic) {
            props.alarm.addInsufficientDataAction(new aws_cloudwatch_actions_1.SnsAction(this.onInsufficientDataTopic));
        }
    }
}
exports.SnsAlarmActionStrategy = SnsAlarmActionStrategy;
_a = JSII_RTTI_SYMBOL_1;
SnsAlarmActionStrategy[_a] = { fqn: "cdk-monitoring-constructs.SnsAlarmActionStrategy", version: "8.1.0" };
