// cdk-monitoring-constructs/lib/monitoring/aws-synthetics/SyntheticsCanaryMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SyntheticsCanaryMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const index_1 = require("../../common/index");
const MetricNamespace = "CloudWatchSynthetics";
class SyntheticsCanaryMetricFactory extends index_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.canary = props.canary;
        this.rateComputationMethod =
            props.rateComputationMethod ?? index_1.RateComputationMethod.AVERAGE;
        this.dimensionsMap = {
            CanaryName: props.canary.canaryName,
        };
    }
    metricLatencyAverageInMillis() {
        return this.metricFactory.adaptMetric(this.canary.metricDuration({
            label: "Average",
            statistic: index_1.MetricStatistic.AVERAGE,
            region: this.region,
            account: this.account,
        }));
    }
    metricSuccessInPercent() {
        return this.metricFactory.adaptMetric(this.canary.metricSuccessPercent({
            label: "Success Rate",
            statistic: index_1.MetricStatistic.AVERAGE,
            region: this.region,
            account: this.account,
        }));
    }
    metric4xxErrorCount() {
        return this.metricFactory.createMetric("4xx", index_1.MetricStatistic.SUM, "4xx", this.dimensionsMap, undefined, MetricNamespace, undefined, this.region, this.account);
    }
    metric4xxErrorRate() {
        return this.metricFactory.toRate(this.metric4xxErrorCount(), this.rateComputationMethod, false, "errors");
    }
    metric5xxFaultCount() {
        return this.metricFactory.createMetric("5xx", index_1.MetricStatistic.SUM, "5xx", this.dimensionsMap, undefined, MetricNamespace, undefined, this.region, this.account);
    }
    metric5xxFaultRate() {
        return this.metricFactory.toRate(this.metric5xxFaultCount(), this.rateComputationMethod, false, "faults");
    }
}
exports.SyntheticsCanaryMetricFactory = SyntheticsCanaryMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
SyntheticsCanaryMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.SyntheticsCanaryMetricFactory", version: "8.1.0" };
