// @cdklabs/generative-ai-cdk-constructs/lib/cdk-lib/bedrock/custom-resource-provider.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BedrockCRProvider = void 0;
/**
 *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
 *  with the License. A copy of the License is located at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 *  or in the 'license' file accompanying this file. This file is distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES
 *  OR CONDITIONS OF ANY KIND, express or implied. See the License for the specific language governing permissions
 *  and limitations under the License.
 */
const path = require("path");
const lambda = require("aws-cdk-lib/aws-lambda");
const custom_resource_provider_helper_1 = require("../../common/helpers/custom-resource-provider-helper");
/**
 * The custom resource provider for Bedrock.
 *
 * @internal This is an internal core function and should not be called directly by Solutions Constructs clients.
 */
exports.BedrockCRProvider = (0, custom_resource_provider_helper_1.buildCustomResourceProvider)({
    providerName: 'BedrockCRProvider',
    codePath: path.join(__dirname, '../../../lambda/bedrock-custom-resources'),
    handler: 'custom_resources.on_event',
    runtime: lambda.Runtime.PYTHON_3_12,
});
