// cdk-monitoring-constructs/lib/monitoring/aws-step-functions/StepFunctionMonitoring.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.StepFunctionMonitoring = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
const StepFunctionMetricFactory_1 = require("./StepFunctionMetricFactory");
const common_1 = require("../../common");
const dashboard_1 = require("../../dashboard");
class StepFunctionMonitoring extends common_1.Monitoring {
    constructor(scope, props) {
        super(scope, props);
        const namedConstruct = props.stateMachine;
        const fallbackConstructName = namedConstruct.stateMachineArn;
        const stateMachineArn = namedConstruct.stateMachineArn;
        const namingStrategy = new dashboard_1.MonitoringNamingStrategy({
            ...props,
            namedConstruct,
            fallbackConstructName,
        });
        this.title = namingStrategy.resolveHumanReadableName();
        this.stateMachineUrl = scope
            .createAwsConsoleUrlFactory()
            .getStateMachineUrl(stateMachineArn);
        const alarmFactory = this.createAlarmFactory(namingStrategy.resolveAlarmFriendlyName());
        this.errorAlarmFactory = new common_1.ErrorAlarmFactory(alarmFactory);
        this.durationAlarmFactory = new common_1.LatencyAlarmFactory(alarmFactory);
        this.taskHealthAlarmFactory = new common_1.TaskHealthAlarmFactory(alarmFactory);
        this.durationAnnotations = [];
        this.errorCountAnnotations = [];
        this.errorRateAnnotations = [];
        const metricFactory = new StepFunctionMetricFactory_1.StepFunctionMetricFactory(scope.createMetricFactory(), props);
        this.p50DurationMetric = metricFactory.metricExecutionTimeP50InMillis();
        this.p90DurationMetric = metricFactory.metricExecutionTimeP90InMillis();
        this.p99DurationMetric = metricFactory.metricExecutionTimeP99InMillis();
        this.startedExecutionsMetric = metricFactory.metricExecutionsStarted();
        this.succeededExecutionsMetric = metricFactory.metricExecutionsSucceeded();
        this.failedExecutionsMetric = metricFactory.metricExecutionsFailed();
        this.failedExecutionRateMetric = metricFactory.metricExecutionsFailedRate();
        this.abortedExecutionsMetric = metricFactory.metricExecutionsAborted();
        this.throttledExecutionsMetric = metricFactory.metricExecutionThrottled();
        this.timedOutExecutionsMetrics = metricFactory.metricExecutionsTimedOut();
        for (const disambiguator in props.addDurationP50Alarm) {
            const alarmProps = props.addDurationP50Alarm[disambiguator];
            const createdAlarm = this.durationAlarmFactory.addDurationAlarm(this.p50DurationMetric, common_1.LatencyType.P50, alarmProps, disambiguator);
            this.durationAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        for (const disambiguator in props.addDurationP90Alarm) {
            const alarmProps = props.addDurationP90Alarm[disambiguator];
            const createdAlarm = this.durationAlarmFactory.addDurationAlarm(this.p90DurationMetric, common_1.LatencyType.P90, alarmProps, disambiguator);
            this.durationAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        for (const disambiguator in props.addDurationP99Alarm) {
            const alarmProps = props.addDurationP99Alarm[disambiguator];
            const createdAlarm = this.durationAlarmFactory.addDurationAlarm(this.p99DurationMetric, common_1.LatencyType.P99, alarmProps, disambiguator);
            this.durationAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        for (const disambiguator in props.addFailedExecutionCountAlarm) {
            const alarmProps = props.addFailedExecutionCountAlarm[disambiguator];
            const createdAlarm = this.errorAlarmFactory.addErrorCountAlarm(this.failedExecutionsMetric, common_1.ErrorType.FAILURE, alarmProps, disambiguator);
            this.errorCountAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        for (const disambiguator in props.addFailedExecutionRateAlarm) {
            const alarmProps = props.addFailedExecutionRateAlarm[disambiguator];
            const createdAlarm = this.errorAlarmFactory.addErrorRateAlarm(this.failedExecutionRateMetric, common_1.ErrorType.FAILURE, alarmProps, disambiguator);
            this.errorRateAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        for (const disambiguator in props.addAbortedExecutionCountAlarm) {
            const alarmProps = props.addAbortedExecutionCountAlarm[disambiguator];
            const createdAlarm = this.errorAlarmFactory.addErrorCountAlarm(this.abortedExecutionsMetric, common_1.ErrorType.ABORTED, alarmProps, disambiguator);
            this.errorCountAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        for (const disambiguator in props.addThrottledExecutionCountAlarm) {
            const alarmProps = props.addThrottledExecutionCountAlarm[disambiguator];
            const createdAlarm = this.errorAlarmFactory.addErrorCountAlarm(this.throttledExecutionsMetric, common_1.ErrorType.THROTTLED, alarmProps, disambiguator);
            this.errorCountAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        for (const disambiguator in props.addTimedOutExecutionCountAlarm) {
            const alarmProps = props.addTimedOutExecutionCountAlarm[disambiguator];
            const createdAlarm = this.errorAlarmFactory.addErrorCountAlarm(this.timedOutExecutionsMetrics, common_1.ErrorType.TIMED_OUT, alarmProps, disambiguator);
            this.errorCountAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        for (const disambiguator in props.addMinStartedExecutionCountAlarm) {
            const alarmProps = props.addMinStartedExecutionCountAlarm[disambiguator];
            const createdAlarm = this.taskHealthAlarmFactory.addMinRunningTaskCountAlarm(this.startedExecutionsMetric, alarmProps, disambiguator);
            this.errorCountAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        props.useCreatedAlarms?.consume(this.createdAlarms());
    }
    summaryWidgets() {
        return [
            // Title
            new dashboard_1.MonitoringHeaderWidget({
                family: "State Machine",
                title: this.title,
                goToLinkUrl: this.stateMachineUrl,
            }),
            // Duration
            new aws_cloudwatch_1.GraphWidget({
                width: common_1.HalfWidth,
                height: common_1.DefaultSummaryWidgetHeight,
                title: "Duration",
                left: [
                    this.p50DurationMetric,
                    this.p90DurationMetric,
                    this.p99DurationMetric,
                ],
                leftYAxis: common_1.TimeAxisMillisFromZero,
                leftAnnotations: this.durationAnnotations,
            }),
            // Statuses
            new aws_cloudwatch_1.GraphWidget({
                width: common_1.HalfWidth,
                height: common_1.DefaultSummaryWidgetHeight,
                title: "Executions",
                left: [
                    this.startedExecutionsMetric,
                    this.succeededExecutionsMetric,
                    this.failedExecutionsMetric,
                    this.abortedExecutionsMetric,
                    this.throttledExecutionsMetric,
                    this.timedOutExecutionsMetrics,
                ],
                leftYAxis: common_1.CountAxisFromZero,
                leftAnnotations: this.errorCountAnnotations,
            }),
        ];
    }
    widgets() {
        return [
            // Title
            new dashboard_1.MonitoringHeaderWidget({
                family: "State Machine",
                title: this.title,
                goToLinkUrl: this.stateMachineUrl,
            }),
            // Duration
            new aws_cloudwatch_1.GraphWidget({
                width: common_1.QuarterWidth,
                height: common_1.DefaultGraphWidgetHeight,
                title: "Duration",
                left: [
                    this.p50DurationMetric,
                    this.p90DurationMetric,
                    this.p99DurationMetric,
                ],
                leftYAxis: common_1.TimeAxisMillisFromZero,
                leftAnnotations: this.durationAnnotations,
            }),
            // Statuses
            new aws_cloudwatch_1.GraphWidget({
                width: common_1.HalfWidth,
                height: common_1.DefaultGraphWidgetHeight,
                title: "Executions",
                left: [
                    this.startedExecutionsMetric,
                    this.succeededExecutionsMetric,
                    this.failedExecutionsMetric,
                    this.abortedExecutionsMetric,
                    this.throttledExecutionsMetric,
                    this.timedOutExecutionsMetrics,
                ],
                leftYAxis: common_1.CountAxisFromZero,
                leftAnnotations: this.errorCountAnnotations,
            }),
            // Fault Rate
            new aws_cloudwatch_1.GraphWidget({
                width: common_1.QuarterWidth,
                height: common_1.DefaultGraphWidgetHeight,
                title: "Errors (rate)",
                left: [this.failedExecutionRateMetric],
                leftAnnotations: this.errorRateAnnotations,
            }),
        ];
    }
}
exports.StepFunctionMonitoring = StepFunctionMonitoring;
_a = JSII_RTTI_SYMBOL_1;
StepFunctionMonitoring[_a] = { fqn: "cdk-monitoring-constructs.StepFunctionMonitoring", version: "8.1.0" };
