// cdk-monitoring-constructs/lib/monitoring/aws-s3/S3BucketMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.S3BucketMetricFactory = exports.StorageType = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
const Namespace = "AWS/S3";
var StorageType;
(function (StorageType) {
    StorageType["DEEP_ARCHIVE_OBJECT_OVERHEAD"] = "DeepArchiveObjectOverhead";
    StorageType["DEEP_ARCHIVE_S3_OBJECT_OVERHEAD"] = "DeepArchiveS3ObjectOverhead";
    StorageType["DEEP_ARCHIVE_STAGING_STORAGE"] = "DeepArchiveStagingStorage";
    StorageType["DEEP_ARCHIVE_STORAGE"] = "DeepArchiveStorage";
    StorageType["GLACIER_OBJECT_OVERHEAD"] = "GlacierObjectOverhead";
    StorageType["GLACIER_S3_OBJECT_OVERHEAD"] = "GlacierS3ObjectOverhead";
    StorageType["GLACIER_STAGING_STORAGE"] = "GlacierStagingStorage";
    StorageType["GLACIER_STORAGE"] = "GlacierStorage";
    StorageType["INTELLIGENT_TIERING_FA_STORAGE"] = "IntelligentTieringFAStorage";
    StorageType["INTELLIGENT_TIERING_IA_STORAGE"] = "IntelligentTieringIAStorage";
    StorageType["ONE_ZONE_IA_SIZE_OVERHEAD"] = "OneZoneIASizeOverhead";
    StorageType["ONE_ZONE_IA_STORAGE"] = "OneZoneIAStorage";
    StorageType["REDUCED_REDUNDANCY_STORAGE"] = "ReducedRedundancyStorage";
    StorageType["STANDARD_IA_SIZE_OVERHEAD"] = "StandardIASizeOverhead";
    StorageType["STANDARD_IA_STORAGE"] = "StandardIAStorage";
    StorageType["STANDARD_STORAGE"] = "StandardStorage";
})(StorageType = exports.StorageType || (exports.StorageType = {}));
class S3BucketMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.props = props;
    }
    metricBucketSizeBytes() {
        return this.metricFactory.createMetric("BucketSizeBytes", common_1.MetricStatistic.AVERAGE, "BucketSizeBytes", {
            BucketName: this.props.bucket.bucketName,
            StorageType: this.props.storageType ?? StorageType.STANDARD_STORAGE,
        }, undefined, Namespace, undefined, this.region, this.account);
    }
    metricNumberOfObjects() {
        return this.metricFactory.createMetric("NumberOfObjects", common_1.MetricStatistic.AVERAGE, "NumberOfObjects", {
            BucketName: this.props.bucket.bucketName,
            StorageType: "AllStorageTypes",
        }, undefined, Namespace, undefined, this.region, this.account);
    }
}
exports.S3BucketMetricFactory = S3BucketMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
S3BucketMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.S3BucketMetricFactory", version: "8.1.0" };
