// cdk-monitoring-constructs/lib/monitoring/aws-ecs-patterns/BaseServiceMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseServiceMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
const EcsNamespace = "AWS/ECS";
const EcsContainerInsightsNamespace = "ECS/ContainerInsights";
/**
 * Metric factory for a base service (parent class for e.g. Fargate and EC2 services).
 */
class BaseServiceMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.dimensionsMap = {
            ClusterName: props.service.cluster.clusterName,
            ServiceName: props.service.serviceName,
        };
        this.service = props.service;
    }
    metricClusterCpuUtilisationInPercent() {
        return this.metricFactory.createMetric("CPUUtilization", common_1.MetricStatistic.AVERAGE, "Cluster CPU Utilization", this.dimensionsMap, undefined, EcsNamespace, undefined, this.region, this.account);
    }
    metricClusterMemoryUtilisationInPercent() {
        return this.metricFactory.createMetric("MemoryUtilization", common_1.MetricStatistic.AVERAGE, "Cluster Memory Utilization", this.dimensionsMap, undefined, EcsNamespace, undefined, this.region, this.account);
    }
    metricRunningTaskCount() {
        return this.metricFactory.createMetric("RunningTaskCount", common_1.MetricStatistic.AVERAGE, "Running Tasks", this.dimensionsMap, undefined, EcsContainerInsightsNamespace, undefined, this.region, this.account);
    }
    metricEphemeralStorageReserved() {
        return this.metricFactory.createMetric("EphemeralStorageReserved", common_1.MetricStatistic.MAX, "Ephemeral Storage Reserved", this.dimensionsMap, undefined, EcsContainerInsightsNamespace, undefined, this.region, this.account);
    }
    metricEphemeralStorageUtilized() {
        return this.metricFactory.createMetric("EphemeralStorageUtilized", common_1.MetricStatistic.MAX, "Ephemeral Storage Utilized", this.dimensionsMap, undefined, EcsContainerInsightsNamespace, undefined, this.region, this.account);
    }
    metricEphemeralStorageUsageInPercent() {
        return this.metricFactory.createMetricMath("100 * (used/total)", {
            used: this.metricEphemeralStorageUtilized(),
            total: this.metricEphemeralStorageReserved(),
        }, "Ephemeral Storage Usage");
    }
}
exports.BaseServiceMetricFactory = BaseServiceMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
BaseServiceMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.BaseServiceMetricFactory", version: "8.1.0" };
