// cdk-monitoring-constructs/lib/monitoring/aws-secretsmanager/SecretsManagerSecretMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsManagerSecretMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
class SecretsManagerSecretMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.secret = props.secret;
        this.dimensionsMap = {
            SecretName: props.secret.secretName,
        };
    }
    metricDaysSinceLastChange() {
        return this.metricFactory.createMetric(SecretsManagerSecretMetricFactory.MetricNameDaysSinceLastChange, common_1.MetricStatistic.MAX, "Days", this.dimensionsMap, undefined, SecretsManagerSecretMetricFactory.Namespace, undefined, this.region, this.account);
    }
    metricDaysSinceLastRotation() {
        return this.metricFactory.createMetric(SecretsManagerSecretMetricFactory.MetricNameDaysSinceLastRotation, common_1.MetricStatistic.MAX, "Days", this.dimensionsMap, undefined, SecretsManagerSecretMetricFactory.Namespace, undefined, this.region, this.account);
    }
}
exports.SecretsManagerSecretMetricFactory = SecretsManagerSecretMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
SecretsManagerSecretMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.SecretsManagerSecretMetricFactory", version: "8.1.0" };
SecretsManagerSecretMetricFactory.Namespace = "SecretsManager";
SecretsManagerSecretMetricFactory.MetricNameDaysSinceLastChange = "DaysSinceLastChange";
SecretsManagerSecretMetricFactory.MetricNameDaysSinceLastRotation = "DaysSinceLastRotation";
