// cdk-monitoring-constructs/lib/monitoring/aws-secretsmanager/SecretsManagerMetricsPublisher.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsManagerMetricsPublisher = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const path = require("path");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_events_1 = require("aws-cdk-lib/aws-events");
const aws_events_targets_1 = require("aws-cdk-lib/aws-events-targets");
const aws_iam_1 = require("aws-cdk-lib/aws-iam");
const aws_lambda_1 = require("aws-cdk-lib/aws-lambda");
const aws_logs_1 = require("aws-cdk-lib/aws-logs");
const constructs_1 = require("constructs");
const SecretsManagerSecretMetricFactory_1 = require("./SecretsManagerSecretMetricFactory");
class SecretsManagerMetricsPublisher extends constructs_1.Construct {
    constructor(scope) {
        super(scope, "SecretsManagerMetricsPublisher");
        this.lambda = new aws_lambda_1.Function(this, "Lambda", {
            code: aws_lambda_1.Code.fromAsset(path.join(__dirname, "..", "..", "..", "assets", "SecretsManagerMetricsPublisher")),
            description: "Custom metrics publisher for SecretsManager Secrets (cdk-monitoring-constructs)",
            handler: "index.handler",
            memorySize: 128,
            runtime: aws_lambda_1.Runtime.NODEJS_18_X,
            timeout: aws_cdk_lib_1.Duration.seconds(60),
            logRetention: aws_logs_1.RetentionDays.ONE_DAY,
        });
        this.lambda.addToRolePolicy(new aws_iam_1.PolicyStatement({
            effect: aws_iam_1.Effect.ALLOW,
            actions: ["secretsmanager:DescribeSecret"],
            resources: ["*"],
        }));
        this.lambda.addToRolePolicy(new aws_iam_1.PolicyStatement({
            effect: aws_iam_1.Effect.ALLOW,
            actions: ["cloudwatch:PutMetricData"],
            resources: ["*"],
            conditions: {
                StringEquals: {
                    "cloudwatch:namespace": SecretsManagerSecretMetricFactory_1.SecretsManagerSecretMetricFactory.Namespace,
                },
            },
        }));
    }
    static getInstance(scope) {
        const key = aws_cdk_lib_1.Names.nodeUniqueId(scope.node);
        let instance = SecretsManagerMetricsPublisher.instances[key];
        if (!instance) {
            instance = new SecretsManagerMetricsPublisher(scope);
            SecretsManagerMetricsPublisher.instances[key] = instance;
        }
        return instance;
    }
    addSecret(secret) {
        // run 1/hr so alarms can recover automatically
        const rule = new aws_events_1.Rule(this, `RuleFor${aws_cdk_lib_1.Names.nodeUniqueId(secret.node)}`, {
            schedule: aws_events_1.Schedule.cron({
                minute: "0",
            }),
        });
        rule.addTarget(new aws_events_targets_1.LambdaFunction(this.lambda, {
            event: aws_events_1.RuleTargetInput.fromObject({
                secretId: secret.secretArn,
            }),
        }));
    }
}
exports.SecretsManagerMetricsPublisher = SecretsManagerMetricsPublisher;
_a = JSII_RTTI_SYMBOL_1;
SecretsManagerMetricsPublisher[_a] = { fqn: "cdk-monitoring-constructs.SecretsManagerMetricsPublisher", version: "8.1.0" };
SecretsManagerMetricsPublisher.instances = {};
