// cdk-monitoring-constructs/lib/monitoring/aws-secretsmanager/SecretsManagerMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsManagerMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const common_1 = require("../../common");
const CLASS = "None";
const DEFAULT_METRIC_PERIOD = aws_cdk_lib_1.Duration.hours(1);
const METRICNAMESECRETCOUNT = "ResourceCount";
const NAMESPACE = "AWS/SecretsManager";
const RESOURCE = "SecretCount";
const SERVICE = "Secrets Manager";
const TYPE = "Resource";
class SecretsManagerMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
    }
    metricSecretCount() {
        const dimensionsMap = {
            Class: CLASS,
            Resource: RESOURCE,
            Service: SERVICE,
            Type: TYPE,
        };
        return this.metricFactory.createMetric(METRICNAMESECRETCOUNT, common_1.MetricStatistic.AVERAGE, "Count", dimensionsMap, undefined, NAMESPACE, DEFAULT_METRIC_PERIOD, this.region, this.account);
    }
}
exports.SecretsManagerMetricFactory = SecretsManagerMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
SecretsManagerMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.SecretsManagerMetricFactory", version: "8.1.0" };
