// cdk-monitoring-constructs/lib/common/monitoring/Monitoring.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.Monitoring = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const dashboard_1 = require("../../dashboard");
/**
 * An independent unit of monitoring. This is the base for all monitoring classes with alarm support.
 */
class Monitoring {
    constructor(scope, props) {
        this.scope = scope;
        this.alarms = [];
        this.localAlarmNamePrefixOverride = props?.localAlarmNamePrefixOverride;
    }
    /**
     * Creates a new widget factory.
     */
    createWidgetFactory() {
        return this.scope.createWidgetFactory();
    }
    /**
     * Creates a new metric factory.
     */
    createMetricFactory() {
        return this.scope.createMetricFactory();
    }
    /**
     * Creates a new alarm factory.
     * Alarms created will be named with the given prefix, unless a local name override is present.
     * @param alarmNamePrefix alarm name prefix
     */
    createAlarmFactory(alarmNamePrefix) {
        return this.scope.createAlarmFactory(this.localAlarmNamePrefixOverride ?? alarmNamePrefix);
    }
    /**
     * Adds an alarm.
     * @param alarm alarm to add
     */
    addAlarm(alarm) {
        this.alarms.push(alarm);
    }
    /**
     * Returns all the alarms created.
     */
    createdAlarms() {
        return this.alarms;
    }
    /**
     * Returns widgets for all alarms. These can go to runbook or to service dashboard.
     */
    alarmWidgets() {
        return this.createdAlarms().map((alarm) => this.createWidgetFactory().createAlarmDetailWidget(alarm));
    }
    /**
     * Returns widgets to be placed on the summary dashboard.
     *
     * @default - no widgets.
     */
    summaryWidgets() {
        return [];
    }
    widgetsForDashboard(name) {
        switch (name) {
            case dashboard_1.DefaultDashboards.SUMMARY:
                return this.summaryWidgets();
            case dashboard_1.DefaultDashboards.DETAIL:
                return this.widgets();
            case dashboard_1.DefaultDashboards.ALARMS:
                return this.alarmWidgets();
            default:
                return [];
        }
    }
}
exports.Monitoring = Monitoring;
_a = JSII_RTTI_SYMBOL_1;
Monitoring[_a] = { fqn: "cdk-monitoring-constructs.Monitoring", version: "8.1.0" };
