// cdk-monitoring-constructs/lib/common/monitoring/alarms/AgeAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AgeAlarmFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
class AgeAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addDaysToExpiryAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.LESS_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.minDaysToExpiry,
            alarmNameSuffix: "DaysToExpiry",
            alarmDescription: "Number of days until expiration is too low.",
        });
    }
    addIteratorMaxAgeAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxAgeInMillis,
            alarmNameSuffix: "Iterator-Age-Max",
            alarmDescription: "Iterator Max Age is too high.",
            // Dedupe all iterator max age to the same ticket
            alarmDedupeStringSuffix: "AnyIteratorMaxAge",
        });
    }
    addDaysSinceUpdateAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxDaysSinceUpdate,
            alarmNameSuffix: "DaysSinceUpdate",
            alarmDescription: "Number of days since update is too high.",
        });
    }
}
exports.AgeAlarmFactory = AgeAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
AgeAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.AgeAlarmFactory", version: "8.1.0" };
