// cdk-monitoring-constructs/lib/common/monitoring/alarms/AuroraAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AuroraAlarmFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
class AuroraAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addMaxServerlessDatabaseCapacity(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxServerlessDatabaseCapacity,
            alarmNameSuffix: "Serverless-Database-Capacity-High",
            alarmDescription: "Serverless Database Capacity usage is too high.",
        });
    }
}
exports.AuroraAlarmFactory = AuroraAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
AuroraAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.AuroraAlarmFactory", version: "8.1.0" };
