// cdk-monitoring-constructs/lib/common/monitoring/alarms/ConnectionAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ConnectionAlarmFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
class ConnectionAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addMinConnectionCountAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.LESS_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.minConnectionCount,
            alarmNameSuffix: "Connection-Count-Low",
            alarmDescription: `Number of connections is too low.`,
        });
    }
    addMaxConnectionCountAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxConnectionCount,
            alarmNameSuffix: "Connection-Count-High",
            alarmDescription: `Number of connections is too high.`,
        });
    }
}
exports.ConnectionAlarmFactory = ConnectionAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
ConnectionAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.ConnectionAlarmFactory", version: "8.1.0" };
