// cdk-monitoring-constructs/lib/common/monitoring/alarms/CustomAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CustomAlarmFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
class CustomAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addCustomAlarm(metric, alarmNameSuffix, disambiguator, props) {
        return this.alarmFactory.addAlarm(metric, {
            ...props,
            disambiguator,
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            threshold: props.threshold,
            comparisonOperator: props.comparisonOperatorOverride ?? props.comparisonOperator,
            alarmDedupeStringSuffix: props.dedupeString,
            alarmNameSuffix,
            alarmDescription: props.additionalDescription ??
                `Threshold of ${props.threshold} has been breached.`,
        });
    }
}
exports.CustomAlarmFactory = CustomAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
CustomAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.CustomAlarmFactory", version: "8.1.0" };
