// cdk-monitoring-constructs/lib/common/monitoring/alarms/DynamoAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DynamoAlarmFactory = exports.CapacityType = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
var CapacityType;
(function (CapacityType) {
    CapacityType["READ"] = "Read";
    CapacityType["WRITE"] = "Write";
})(CapacityType = exports.CapacityType || (exports.CapacityType = {}));
class DynamoAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addConsumedCapacityAlarm(metric, capacityType, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.NOT_BREACHING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxConsumedCapacityUnits,
            alarmNameSuffix: `${capacityType}-Consumed-Capacity`,
            // we will dedupe any kind of error to the same ticket
            alarmDedupeStringSuffix: "ConsumedCapacity",
            alarmDescription: `${capacityType} consumed capacity is too high.`,
        });
    }
    addThrottledEventsAlarm(metric, capacityType, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.NOT_BREACHING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxThrottledEventsThreshold,
            alarmNameSuffix: `${capacityType}-Throttled-Events`,
            // we will dedupe any kind of error to the same ticket
            alarmDedupeStringSuffix: "ThrottledEvents",
            alarmDescription: `${capacityType} throttled events above threshold.`,
        });
    }
}
exports.DynamoAlarmFactory = DynamoAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
DynamoAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.DynamoAlarmFactory", version: "8.1.0" };
