// cdk-monitoring-constructs/lib/common/monitoring/alarms/ErrorAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ErrorAlarmFactory = exports.ErrorType = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
var ErrorType;
(function (ErrorType) {
    ErrorType["FAULT"] = "Fault";
    ErrorType["ERROR"] = "Error";
    ErrorType["SYSTEM_ERROR"] = "SystemError";
    ErrorType["USER_ERROR"] = "UserError";
    ErrorType["FAILURE"] = "Failure";
    ErrorType["ABORTED"] = "Aborted";
    ErrorType["THROTTLED"] = "Throttled";
    ErrorType["TIMED_OUT"] = "TimedOut";
    ErrorType["READ_ERROR"] = "ReadError";
    ErrorType["WRITE_ERROR"] = "WriteError";
    ErrorType["EXPIRED"] = "Expired";
    ErrorType["KILLED"] = "Killed";
    ErrorType["BLOCKED"] = "Blocked";
})(ErrorType = exports.ErrorType || (exports.ErrorType = {}));
class ErrorAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addErrorCountAlarm(metric, errorType, props, disambiguator) {
        const alarmNameSuffix = `${errorType}-Count`;
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.NOT_BREACHING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxErrorCount,
            alarmNameSuffix,
            // we will dedupe any kind of error to the same ticket
            alarmDedupeStringSuffix: this.alarmFactory.shouldUseDefaultDedupeForError
                ? "AnyError"
                : alarmNameSuffix,
            alarmDescription: `${errorType} count is too high.`,
        });
    }
    addErrorRateAlarm(metric, errorType, props, disambiguator) {
        const alarmNameSuffix = `${errorType}-Rate`;
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.NOT_BREACHING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxErrorRate,
            alarmNameSuffix,
            // we will dedupe any kind of error to the same ticket
            alarmDedupeStringSuffix: this.alarmFactory.shouldUseDefaultDedupeForError
                ? "AnyError"
                : alarmNameSuffix,
            alarmDescription: `${errorType} rate is too high.`,
        });
    }
}
exports.ErrorAlarmFactory = ErrorAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
ErrorAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.ErrorAlarmFactory", version: "8.1.0" };
