// cdk-monitoring-constructs/lib/common/monitoring/alarms/KinesisDataAnalyticsAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KinesisDataAnalyticsAlarmFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
class KinesisDataAnalyticsAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addDowntimeAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.BREACHING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxDowntimeInMillis,
            alarmNameSuffix: "Downtime",
            alarmDescription: "Application has too much downtime",
            // we will dedupe any kind of message count issue to the same ticket
            alarmDedupeStringSuffix: "KDADowntimeAlarm",
        });
    }
    addFullRestartAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.BREACHING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxFullRestartCount,
            alarmNameSuffix: "FullRestart",
            alarmDescription: "Last submitted job is restarting more than usual",
            alarmDedupeStringSuffix: "KDAFullRestartAlarm",
        });
    }
}
exports.KinesisDataAnalyticsAlarmFactory = KinesisDataAnalyticsAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
KinesisDataAnalyticsAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.KinesisDataAnalyticsAlarmFactory", version: "8.1.0" };
