// cdk-monitoring-constructs/lib/common/monitoring/alarms/SecretsManagerAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SecretsManagerAlarmFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
const NUMBER_OF_DATAPOINTS = 1;
class SecretsManagerAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addMinSecretCountAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            datapointsToAlarm: props.datapointsToAlarm ?? NUMBER_OF_DATAPOINTS,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.LESS_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.minSecretCount,
            alarmNameSuffix: "Secrets-Count-Min",
            alarmDescription: "Number of secrets is too low.",
        });
    }
    addMaxSecretCountAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            datapointsToAlarm: props.datapointsToAlarm ?? NUMBER_OF_DATAPOINTS,
            ...props,
            disambiguator,
            threshold: props.maxSecretCount,
            alarmNameSuffix: "Secrets-Count-Max",
            alarmDescription: "Number of secrets is too high.",
        });
    }
    addChangeInSecretCountAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            ...props,
            disambiguator,
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            threshold: props.requiredSecretCount,
            comparisonOperator: this.getComparisonOperator(props),
            datapointsToAlarm: props.datapointsToAlarm ?? NUMBER_OF_DATAPOINTS,
            alarmNameSuffix: "Secrets-Count-Change",
            alarmDescription: this.getDefaultDescription(props),
        });
    }
    getDefaultDescription(props) {
        if (props.alarmWhenIncreased && props.alarmWhenDecreased) {
            return "Secret count: Secret count has changed.";
        }
        else if (props.alarmWhenIncreased) {
            return "Secret count: Secret count has increased.";
        }
        else if (props.alarmWhenDecreased) {
            return "Secret count: Secret count has decreased.";
        }
        else {
            throw new Error("You need to alarm when the value has increased, decreased, or both.");
        }
    }
    getComparisonOperator(props) {
        if (props.alarmWhenIncreased && props.alarmWhenDecreased) {
            return aws_cloudwatch_1.ComparisonOperator.LESS_THAN_LOWER_OR_GREATER_THAN_UPPER_THRESHOLD;
        }
        else if (props.alarmWhenDecreased) {
            return aws_cloudwatch_1.ComparisonOperator.LESS_THAN_THRESHOLD;
        }
        else if (props.alarmWhenIncreased) {
            return aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD;
        }
        else {
            throw new Error("You need to alarm when the value has increased, decreased, or both.");
        }
    }
}
exports.SecretsManagerAlarmFactory = SecretsManagerAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
SecretsManagerAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.SecretsManagerAlarmFactory", version: "8.1.0" };
