// cdk-monitoring-constructs/lib/common/monitoring/alarms/ThroughputAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ThroughputAlarmFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
class ThroughputAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addMinProcessedBytesAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.NOT_BREACHING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.LESS_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.minProcessedBytes,
            alarmNameSuffix: "Processed-Bytes-Min",
            alarmDescription: `Minimum number of processed bytes is too low.`,
        });
    }
}
exports.ThroughputAlarmFactory = ThroughputAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
ThroughputAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.ThroughputAlarmFactory", version: "8.1.0" };
