// cdk-monitoring-constructs/lib/common/monitoring/alarms/TopicAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.TopicAlarmFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
class TopicAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addMinMessagesPublishedAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.LESS_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.minMessagesPublishedCount,
            alarmNameSuffix: "Topic-Published-Low",
            alarmDescription: `Number of SNS messages published is too low.`,
        });
    }
    addMaxMessagesPublishedAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxMessagesPublishedCount,
            alarmNameSuffix: "Topic-Published-High",
            alarmDescription: `Number of SNS messages published is too high.`,
        });
    }
    addMessageNotificationsFailedAlarm(metric, props, disambiguator) {
        const threshold = props.maxNotificationsFailedCount;
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: threshold,
            alarmNameSuffix: "Topic-Notifications-Failed",
            alarmDescription: `Number of failed sns messages exceeded threshold of ${threshold}.`,
        });
    }
}
exports.TopicAlarmFactory = TopicAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
TopicAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.TopicAlarmFactory", version: "8.1.0" };
