// cdk-monitoring-constructs/lib/common/monitoring/alarms/UsageAlarmFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.UsageAlarmFactory = exports.UsageType = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
var UsageType;
(function (UsageType) {
    UsageType["P50"] = "P50";
    UsageType["P70"] = "P70";
    UsageType["P90"] = "P90";
    UsageType["P99"] = "P99";
    UsageType["P999"] = "P999";
    UsageType["P9999"] = "P9999";
    UsageType["P100"] = "P100";
    UsageType["AVERAGE"] = "Average";
    UsageType["MAX"] = "Maximum";
})(UsageType = exports.UsageType || (exports.UsageType = {}));
class UsageAlarmFactory {
    constructor(alarmFactory) {
        this.alarmFactory = alarmFactory;
    }
    addMaxCountAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxCount,
            alarmNameSuffix: "Max-Usage-Count",
            alarmDescription: "The count is too high.",
        });
    }
    addMinCountAlarm(percentMetric, props, disambiguator) {
        return this.alarmFactory.addAlarm(percentMetric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.LESS_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.minCount,
            alarmNameSuffix: "Min-Usage-Count",
            alarmDescription: "The count is too low.",
        });
    }
    /**
     * @deprecated Use {@link addMaxCountAlarm} instead.
     */
    addMaxUsageCountAlarm(metric, props, disambiguator) {
        return this.alarmFactory.addAlarm(metric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxUsageCount,
            alarmNameSuffix: "Usage-Count",
            alarmDescription: "The count is too high.",
        });
    }
    /**
     * @deprecated Use {@link addMinCountAlarm} instead.
     */
    addMinUsageCountAlarm(percentMetric, props, disambiguator) {
        return this.alarmFactory.addAlarm(percentMetric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.LESS_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.minCount,
            alarmNameSuffix: "Usage-Count",
            alarmDescription: "The count is too low.",
        });
    }
    addMaxCpuUsagePercentAlarm(percentMetric, props, disambiguator, usageType, additionalAlarmNameSuffix) {
        const alarmNameSuffix = [
            usageType,
            "CPU-Usage",
            additionalAlarmNameSuffix,
        ]
            .filter((i) => i !== undefined)
            .join("-");
        return this.alarmFactory.addAlarm(percentMetric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxUsagePercent,
            alarmNameSuffix,
            alarmDescription: "The CPU usage is too high.",
        });
    }
    addMaxMasterCpuUsagePercentAlarm(percentMetric, props, disambiguator) {
        return this.alarmFactory.addAlarm(percentMetric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxUsagePercent,
            alarmNameSuffix: "Master-CPU-Usage",
            alarmDescription: "The master CPU usage is too high.",
        });
    }
    addMaxMemoryUsagePercentAlarm(percentMetric, props, disambiguator) {
        return this.addMemoryUsagePercentAlarm(percentMetric, props, UsageType.MAX, disambiguator);
    }
    addMemoryUsagePercentAlarm(percentMetric, props, usageType, disambiguator) {
        const alarmNameSuffix = usageType === UsageType.MAX
            ? "Memory-Usage"
            : `${usageType}-Memory-Usage`;
        return this.alarmFactory.addAlarm(percentMetric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxUsagePercent,
            alarmNameSuffix,
            alarmDescription: "The memory usage is too high.",
        });
    }
    addMaxMasterMemoryUsagePercentAlarm(percentMetric, props, disambiguator) {
        return this.alarmFactory.addAlarm(percentMetric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxUsagePercent,
            alarmNameSuffix: "Master-Memory-Usage",
            alarmDescription: "The master memory usage is too high.",
        });
    }
    addMaxDiskUsagePercentAlarm(percentMetric, props, disambiguator) {
        return this.alarmFactory.addAlarm(percentMetric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxUsagePercent,
            alarmNameSuffix: "Disk-Usage",
            alarmDescription: "The disk usage is too high.",
        });
    }
    addMaxHeapMemoryAfterGCUsagePercentAlarm(percentMetric, props, disambiguator) {
        return this.alarmFactory.addAlarm(percentMetric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxUsagePercent,
            alarmNameSuffix: "HeapMemoryAfterGC-Usage",
            alarmDescription: "The heap memory after GC usage is too high.",
        });
    }
    addMaxFileDescriptorPercentAlarm(percentMetric, props, disambiguator) {
        return this.alarmFactory.addAlarm(percentMetric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxUsagePercent,
            alarmNameSuffix: "File-Descriptor-Usage",
            alarmDescription: "The file descriptor usage is too high.",
        });
    }
    addMaxThreadCountUsageAlarm(percentMetric, props, disambiguator) {
        return this.alarmFactory.addAlarm(percentMetric, {
            treatMissingData: props.treatMissingDataOverride ?? aws_cloudwatch_1.TreatMissingData.MISSING,
            comparisonOperator: props.comparisonOperatorOverride ??
                aws_cloudwatch_1.ComparisonOperator.GREATER_THAN_THRESHOLD,
            ...props,
            disambiguator,
            threshold: props.maxUsageCount,
            alarmNameSuffix: "Thread-Count",
            alarmDescription: "The thread count is too high.",
        });
    }
}
exports.UsageAlarmFactory = UsageAlarmFactory;
_a = JSII_RTTI_SYMBOL_1;
UsageAlarmFactory[_a] = { fqn: "cdk-monitoring-constructs.UsageAlarmFactory", version: "8.1.0" };
