// cdk-monitoring-constructs/lib/monitoring/aws-kinesisanalytics/KinesisDataAnalyticsMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KinesisDataAnalyticsMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
/**
 * @see https://docs.aws.amazon.com/kinesisanalytics/latest/java/metrics-dimensions.html
 */
class KinesisDataAnalyticsMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.dimensionsMap = {
            Application: props.application,
        };
    }
    metricKPUsCount() {
        return this.metric({
            name: "KPUs",
            description: "Kinesis Processing Units",
        });
    }
    metricDowntimeMs() {
        return this.metric({
            name: "downtime",
            description: "Downtime",
        });
    }
    metricUptimeMs() {
        return this.metric({
            name: "uptime",
            description: "Uptime",
        });
    }
    metricFullRestartsCount() {
        return this.metric({
            name: "fullRestarts",
            description: "Restarts",
        });
    }
    metricNumberOfFailedCheckpointsCount() {
        return this.metric({
            name: "numberOfFailedCheckpoints",
            description: "Failed Checkpoints",
            metricStatistic: common_1.MetricStatistic.SUM,
        });
    }
    metricLastCheckpointDurationMs() {
        return this.metric({
            name: "lastCheckpointDuration",
            description: "Last Checkpoint Duration",
        });
    }
    metricLastCheckpointSizeBytes() {
        return this.metric({
            name: "lastCheckpointSize",
            description: "Last Checkpoint Size",
            metricStatistic: common_1.MetricStatistic.SUM,
        });
    }
    metricCpuUtilizationPercent() {
        return this.metric({
            name: "cpuUtilization",
            description: "CPU Utilization",
        });
    }
    metricHeapMemoryUtilizationPercent() {
        return this.metric({
            name: "heapMemoryUtilization",
            description: "Heap Memory Utilization",
        });
    }
    metricOldGenerationGCTimeMs() {
        return this.metric({
            name: "oldGenerationGCTime",
            description: "GC Time",
        });
    }
    metricOldGenerationGCCount() {
        return this.metric({
            name: "oldGenerationGCCount",
            metricStatistic: common_1.MetricStatistic.N,
            description: "GC Count",
        });
    }
    metric(metricsSpec) {
        return this.metricFactory.createMetric(metricsSpec.name, metricsSpec.metricStatistic ?? common_1.MetricStatistic.AVERAGE, metricsSpec.description, this.dimensionsMap, undefined, "AWS/KinesisAnalytics", undefined, this.region, this.account);
    }
}
exports.KinesisDataAnalyticsMetricFactory = KinesisDataAnalyticsMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
KinesisDataAnalyticsMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.KinesisDataAnalyticsMetricFactory", version: "8.1.0" };
