// cdk-monitoring-constructs/lib/monitoring/aws-kinesisanalytics/KinesisDataAnalyticsMonitoring.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.KinesisDataAnalyticsMonitoring = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
const KinesisDataAnalyticsMetricFactory_1 = require("./KinesisDataAnalyticsMetricFactory");
const common_1 = require("../../common");
const dashboard_1 = require("../../dashboard");
class KinesisDataAnalyticsMonitoring extends common_1.Monitoring {
    constructor(scope, props) {
        super(scope, props);
        const namingStrategy = new dashboard_1.MonitoringNamingStrategy({
            ...props,
            fallbackConstructName: props.application,
        });
        this.title = namingStrategy.resolveHumanReadableName();
        this.kinesisDataAnalyticsUrl = scope
            .createAwsConsoleUrlFactory()
            .getKinesisAnalyticsUrl(props.application);
        const alarmFactory = this.createAlarmFactory(namingStrategy.resolveAlarmFriendlyName());
        this.kdaAlarmFactory = new common_1.KinesisDataAnalyticsAlarmFactory(alarmFactory);
        this.downtimeAnnotations = [];
        this.fullRestartAnnotations = [];
        const metricFactory = new KinesisDataAnalyticsMetricFactory_1.KinesisDataAnalyticsMetricFactory(scope.createMetricFactory(), props);
        this.cpuUtilizationPercentMetric =
            metricFactory.metricCpuUtilizationPercent();
        this.downtimeMsMetric = metricFactory.metricDowntimeMs();
        this.fullRestartsCountMetric = metricFactory.metricFullRestartsCount();
        this.heapMemoryUtilizationPercentMetric =
            metricFactory.metricHeapMemoryUtilizationPercent();
        this.kpusCountMetric = metricFactory.metricKPUsCount();
        this.lastCheckpointDurationMsMetric =
            metricFactory.metricLastCheckpointDurationMs();
        this.lastCheckpointSizeBytesMetric =
            metricFactory.metricLastCheckpointSizeBytes();
        this.numberOfFailedCheckpointsCountMetric =
            metricFactory.metricNumberOfFailedCheckpointsCount();
        this.oldGenerationGCCountMetric =
            metricFactory.metricOldGenerationGCCount();
        this.oldGenerationGCTimeMsMetric =
            metricFactory.metricOldGenerationGCTimeMs();
        for (const disambiguator in props.addDowntimeAlarm) {
            const alarmProps = props.addDowntimeAlarm[disambiguator];
            const createdAlarm = this.kdaAlarmFactory.addDowntimeAlarm(this.downtimeMsMetric, alarmProps, disambiguator);
            this.downtimeAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        for (const disambiguator in props.addFullRestartCountAlarm) {
            const alarmProps = props.addFullRestartCountAlarm[disambiguator];
            const createdAlarm = this.kdaAlarmFactory.addFullRestartAlarm(this.fullRestartsCountMetric, alarmProps, disambiguator);
            this.fullRestartAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        props.useCreatedAlarms?.consume(this.createdAlarms());
    }
    summaryWidgets() {
        return [
            this.createTitleWidget(),
            ...this.createSummaryWidgetRow(common_1.DefaultSummaryWidgetHeight),
        ];
    }
    widgets() {
        return [
            this.createTitleWidget(),
            ...this.createSummaryWidgetRow(common_1.DefaultGraphWidgetHeight),
            ...this.createCheckpointAndGcWidgets(),
        ];
    }
    createTitleWidget() {
        return new dashboard_1.MonitoringHeaderWidget({
            family: "Kinesis Data Analytics",
            title: this.title,
            goToLinkUrl: this.kinesisDataAnalyticsUrl,
        });
    }
    createKPUWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "KPU Usage",
            left: [this.kpusCountMetric],
            leftYAxis: common_1.CountAxisFromZero,
        });
    }
    createResourceUtilizationWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "Resource Utilization",
            left: [
                this.cpuUtilizationPercentMetric,
                this.heapMemoryUtilizationPercentMetric,
            ],
            leftYAxis: common_1.PercentageAxisFromZeroToHundred,
        });
    }
    createDownTimeWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "Down Time",
            left: [this.downtimeMsMetric],
            leftYAxis: common_1.TimeAxisMillisFromZero,
            leftAnnotations: this.downtimeAnnotations,
        });
    }
    createFullRestartsWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "Full Restarts",
            left: [this.fullRestartsCountMetric],
            leftYAxis: common_1.CountAxisFromZero,
            leftAnnotations: this.fullRestartAnnotations,
        });
    }
    createNumberOfFailedCheckpointsWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "Checkpoint Failures",
            left: [this.numberOfFailedCheckpointsCountMetric],
            leftYAxis: common_1.CountAxisFromZero,
        });
    }
    createLastCheckpointDurationWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "Checkpoint Duration",
            left: [this.lastCheckpointDurationMsMetric],
            leftYAxis: common_1.TimeAxisMillisFromZero,
        });
    }
    createLastCheckpointSizeWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "Checkpoint Size",
            left: [this.lastCheckpointSizeBytesMetric],
            leftYAxis: common_1.SizeAxisBytesFromZero,
        });
    }
    createGarbageCollectionWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "Garbage Collection",
            left: [this.oldGenerationGCCountMetric],
            leftYAxis: common_1.CountAxisFromZero,
            right: [this.oldGenerationGCTimeMsMetric],
            rightYAxis: common_1.TimeAxisMillisFromZero,
        });
    }
    createSummaryWidgetRow(height) {
        return [
            // KPUs
            this.createKPUWidget(common_1.QuarterWidth, height),
            // CPU And Heap Usage
            this.createResourceUtilizationWidget(common_1.QuarterWidth, height),
            // Down Time and Up Time
            this.createDownTimeWidget(common_1.QuarterWidth, height),
            // Full Restarts
            this.createFullRestartsWidget(common_1.QuarterWidth, height),
        ];
    }
    createCheckpointAndGcWidgets() {
        return [
            // Checkpointing
            this.createNumberOfFailedCheckpointsWidget(common_1.QuarterWidth, common_1.DefaultGraphWidgetHeight),
            // Checkpoint Duration
            this.createLastCheckpointDurationWidget(common_1.QuarterWidth, common_1.DefaultGraphWidgetHeight),
            // Checkpoint Size
            this.createLastCheckpointSizeWidget(common_1.QuarterWidth, common_1.DefaultGraphWidgetHeight),
            // Garbage Collection
            this.createGarbageCollectionWidget(common_1.QuarterWidth, common_1.DefaultGraphWidgetHeight),
        ];
    }
}
exports.KinesisDataAnalyticsMonitoring = KinesisDataAnalyticsMonitoring;
_a = JSII_RTTI_SYMBOL_1;
KinesisDataAnalyticsMonitoring[_a] = { fqn: "cdk-monitoring-constructs.KinesisDataAnalyticsMonitoring", version: "8.1.0" };
