// cdk-monitoring-constructs/lib/monitoring/aws-ec2/AutoScalingGroupMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoScalingGroupMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
const AutoScalingNamespace = "AWS/AutoScaling";
class AutoScalingGroupMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.dimensionsMap = {
            AutoScalingGroupName: props.autoScalingGroup.autoScalingGroupName,
        };
    }
    /**
     * The minimum size of the Auto Scaling group.
     */
    metricGroupMinSize() {
        return this.metric("GroupMinSize", "Min", common_1.MetricStatistic.AVERAGE);
    }
    /**
     * The maximum size of the Auto Scaling group.
     */
    metricGroupMaxSize() {
        return this.metric("GroupMaxSize", "Max", common_1.MetricStatistic.AVERAGE);
    }
    /**
     * The number of instances that the Auto Scaling group attempts to maintain.
     */
    metricGroupDesiredCapacity() {
        return this.metric("GroupDesiredCapacity", "Desired", common_1.MetricStatistic.AVERAGE);
    }
    /**
       *
       The number of instances that are running as part of the Auto Scaling group.
       This metric does not include instances that are pending or terminating.
       */
    metricGroupInServiceInstances() {
        return this.metric("GroupInServiceInstances", "InService", common_1.MetricStatistic.AVERAGE);
    }
    /**
     * The number of instances that are pending. A pending instance is not yet in service.
     * This metric does not include instances that are in service or terminating.
     */
    metricGroupPendingInstances() {
        return this.metric("GroupPendingInstances", "Pending", common_1.MetricStatistic.AVERAGE);
    }
    /**
     * The number of instances that are in a Standby state.
     * Instances in this state are still running but are not actively in service.
     */
    metricGroupStandbyInstances() {
        return this.metric("GroupStandbyInstances", "Standby", common_1.MetricStatistic.AVERAGE);
    }
    /**
     * The number of instances that are in the process of terminating.
     * This metric does not include instances that are in service or pending.
     */
    metricGroupTerminatingInstances() {
        return this.metric("GroupTerminatingInstances", "Terminating", common_1.MetricStatistic.AVERAGE);
    }
    /**
     * The total number of instances in the Auto Scaling group.
     * This metric identifies the number of instances that are in service, pending, and terminating.
     */
    metricGroupTotalInstances() {
        return this.metric("GroupTotalInstances", "Total", common_1.MetricStatistic.AVERAGE);
    }
    metric(metricName, label, statistic) {
        return this.metricFactory.createMetric(metricName, statistic, label, this.dimensionsMap, undefined, AutoScalingNamespace, undefined, this.region, this.account);
    }
}
exports.AutoScalingGroupMetricFactory = AutoScalingGroupMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
AutoScalingGroupMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.AutoScalingGroupMetricFactory", version: "8.1.0" };
