// cdk-monitoring-constructs/lib/monitoring/aws-ec2/AutoScalingGroupMonitoring.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.AutoScalingGroupMonitoring = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
const AutoScalingGroupMetricFactory_1 = require("./AutoScalingGroupMetricFactory");
const common_1 = require("../../common");
const dashboard_1 = require("../../dashboard");
class AutoScalingGroupMonitoring extends common_1.Monitoring {
    constructor(scope, props) {
        super(scope, props);
        const fallbackConstructName = props.autoScalingGroup.autoScalingGroupName;
        const namingStrategy = new dashboard_1.MonitoringNamingStrategy({
            ...props,
            namedConstruct: props.autoScalingGroup,
            fallbackConstructName,
        });
        this.title = namingStrategy.resolveHumanReadableName();
        const metricFactory = new AutoScalingGroupMetricFactory_1.AutoScalingGroupMetricFactory(scope.createMetricFactory(), props);
        this.groupMinSizeMetric = metricFactory.metricGroupMinSize();
        this.groupMaxSizeMetric = metricFactory.metricGroupMaxSize();
        this.groupDesiredSizeMetric = metricFactory.metricGroupDesiredCapacity();
        this.instancesInServiceMetric =
            metricFactory.metricGroupInServiceInstances();
        this.instancesPendingMetric = metricFactory.metricGroupPendingInstances();
        this.instancesStandbyMetric = metricFactory.metricGroupStandbyInstances();
        this.instancesTerminatingMetric =
            metricFactory.metricGroupTerminatingInstances();
        this.instancesTotalMetric = metricFactory.metricGroupTotalInstances();
    }
    summaryWidgets() {
        return [
            this.createTitleWidget(),
            this.createGroupSizeWidget(common_1.FullWidth, common_1.DefaultSummaryWidgetHeight),
        ];
    }
    widgets() {
        return [
            this.createTitleWidget(),
            this.createGroupSizeWidget(common_1.HalfWidth, common_1.DefaultGraphWidgetHeight),
            this.createGroupStatusWidget(common_1.HalfWidth, common_1.DefaultGraphWidgetHeight),
        ];
    }
    createTitleWidget() {
        return new dashboard_1.MonitoringHeaderWidget({
            family: "Auto Scaling Group",
            title: this.title,
        });
    }
    createGroupSizeWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "Group Size",
            left: [
                this.groupMinSizeMetric,
                this.groupMaxSizeMetric,
                this.groupDesiredSizeMetric,
                this.instancesTotalMetric,
            ],
            leftYAxis: common_1.CountAxisFromZero,
        });
    }
    createGroupStatusWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "Instance States",
            left: [
                this.instancesInServiceMetric,
                this.instancesPendingMetric,
                this.instancesStandbyMetric,
                this.instancesTerminatingMetric,
            ],
            leftYAxis: common_1.CountAxisFromZero,
            stacked: true,
        });
    }
}
exports.AutoScalingGroupMonitoring = AutoScalingGroupMonitoring;
_a = JSII_RTTI_SYMBOL_1;
AutoScalingGroupMonitoring[_a] = { fqn: "cdk-monitoring-constructs.AutoScalingGroupMonitoring", version: "8.1.0" };
