// cdk-monitoring-constructs/lib/monitoring/aws-sns/SnsTopicMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SnsTopicMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
class SnsTopicMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.topic = props.topic;
    }
    metricIncomingMessageCount() {
        return this.metricFactory.adaptMetric(this.topic.metricNumberOfMessagesPublished({
            label: "Incoming",
            region: this.region,
            account: this.account,
        }));
    }
    metricOutgoingMessageCount() {
        return this.metricFactory.adaptMetric(this.topic.metricNumberOfNotificationsDelivered({
            label: "Outgoing",
            region: this.region,
            account: this.account,
        }));
    }
    metricAverageMessageSizeInBytes() {
        return this.metricFactory.adaptMetric(this.topic.metricPublishSize({
            label: "Size",
            region: this.region,
            account: this.account,
        }));
    }
    metricNumberOfNotificationsFailed() {
        return this.metricFactory.adaptMetric(this.topic.metricNumberOfNotificationsFailed({
            label: "Failed",
            region: this.region,
            account: this.account,
        }));
    }
}
exports.SnsTopicMetricFactory = SnsTopicMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
SnsTopicMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.SnsTopicMetricFactory", version: "8.1.0" };
