// cdk-monitoring-constructs/lib/dashboard/widget/HeaderWidget.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.HeaderWidget = exports.HeaderLevel = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
const widget_1 = require("../../common/widget");
var HeaderLevel;
(function (HeaderLevel) {
    HeaderLevel[HeaderLevel["LARGE"] = 0] = "LARGE";
    HeaderLevel[HeaderLevel["MEDIUM"] = 1] = "MEDIUM";
    HeaderLevel[HeaderLevel["SMALL"] = 2] = "SMALL";
})(HeaderLevel = exports.HeaderLevel || (exports.HeaderLevel = {}));
class HeaderWidget extends aws_cloudwatch_1.TextWidget {
    constructor(text, level, description, descriptionHeight) {
        super({
            width: widget_1.FullWidth,
            height: HeaderWidget.calculateHeight(description, descriptionHeight),
            markdown: HeaderWidget.toMarkdown(text, level ?? HeaderLevel.LARGE, description),
        });
    }
    static calculateHeight(description, descriptionHeight) {
        let result = 1;
        if (description) {
            result += descriptionHeight ?? 1;
        }
        return result;
    }
    static toMarkdown(text, level, description) {
        const parts = [this.toHeaderMarkdown(text, level)];
        if (description) {
            parts.push(description);
        }
        return parts.join("\n\n");
    }
    static toHeaderMarkdown(text, level) {
        return "#".repeat(level + 1) + " " + text;
    }
}
exports.HeaderWidget = HeaderWidget;
_a = JSII_RTTI_SYMBOL_1;
HeaderWidget[_a] = { fqn: "cdk-monitoring-constructs.HeaderWidget", version: "8.1.0" };
