// cdk-monitoring-constructs/lib/dashboard/widget/StrictGraphWidget.js
"use strict";
var _a, _b;
Object.defineProperty(exports, "__esModule", { value: true });
exports.DoubleAxisGraphWidget = exports.SingleAxisGraphWidget = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
/**
 * Line graph widget with one axis only (left).
 * If there is just one metric, it will hide the legend to save space.
 * The purpose of this custom class is to make the properties more strict.
 * It will avoid graphs with undefined axis and dimensions.
 */
class SingleAxisGraphWidget extends aws_cloudwatch_1.GraphWidget {
    constructor(props) {
        super(SingleAxisGraphWidget.transformProps(props));
    }
    static transformProps(props) {
        if (props.leftMetrics.length < 1) {
            throw new Error("No metrics defined. Please define at least one metric.");
        }
        const legendPosition = props.leftMetrics.length === 1 ? aws_cloudwatch_1.LegendPosition.HIDDEN : undefined;
        return {
            title: props.title,
            width: props.width,
            height: props.height,
            left: props.leftMetrics,
            leftYAxis: props.leftAxis,
            leftAnnotations: props.leftAnnotations,
            verticalAnnotations: props.verticalAnnotations,
            legendPosition,
        };
    }
}
exports.SingleAxisGraphWidget = SingleAxisGraphWidget;
_a = JSII_RTTI_SYMBOL_1;
SingleAxisGraphWidget[_a] = { fqn: "cdk-monitoring-constructs.SingleAxisGraphWidget", version: "8.1.0" };
/**
 * Line graph widget with both left and right axes.
 * The purpose of this custom class is to make the properties more strict.
 * It will avoid graphs with undefined axes and dimensions.
 */
class DoubleAxisGraphWidget extends aws_cloudwatch_1.GraphWidget {
    constructor(props) {
        super(DoubleAxisGraphWidget.transformProps(props));
    }
    static transformProps(props) {
        if (props.leftMetrics.length < 1) {
            throw new Error("No left metrics defined. Please define at least one metric.");
        }
        if (props.rightMetrics.length < 1) {
            throw new Error("No right metrics defined. Please define at least one metric.");
        }
        return {
            title: props.title,
            width: props.width,
            height: props.height,
            left: props.leftMetrics,
            leftYAxis: props.leftAxis,
            leftAnnotations: props.leftAnnotations,
            right: props.rightMetrics,
            rightYAxis: props.rightAxis,
            rightAnnotations: props.rightAnnotations,
            verticalAnnotations: props.verticalAnnotations,
        };
    }
}
exports.DoubleAxisGraphWidget = DoubleAxisGraphWidget;
_b = JSII_RTTI_SYMBOL_1;
DoubleAxisGraphWidget[_b] = { fqn: "cdk-monitoring-constructs.DoubleAxisGraphWidget", version: "8.1.0" };
