// cdk-monitoring-constructs/lib/monitoring/aws-wafv2/WafV2MetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.WafV2MetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
const MetricNamespace = "AWS/WAFV2";
const AllRulesDimensionValue = "ALL";
/**
 * https://docs.aws.amazon.com/waf/latest/developerguide/monitoring-cloudwatch.html
 */
class WafV2MetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        if (props.acl.scope === "REGIONAL" && !props.region) {
            throw new Error(`region is required if CfnWebACL has "REGIONAL" scope`);
        }
        this.dimensions = {
            Rule: AllRulesDimensionValue,
            WebACL: props.acl.name,
            ...(props.region && { Region: props.region }),
        };
    }
    metricAllowedRequests() {
        return this.metricFactory.createMetric("AllowedRequests", common_1.MetricStatistic.SUM, "Allowed", this.dimensions, undefined, MetricNamespace, undefined, this.region, this.account);
    }
    metricBlockedRequests() {
        return this.metricFactory.createMetric("BlockedRequests", common_1.MetricStatistic.SUM, "Blocked", this.dimensions, undefined, MetricNamespace, undefined, this.region, this.account);
    }
    metricBlockedRequestsRate() {
        return this.metricFactory.createMetricMath("100 * (blocked / (allowed + blocked))", {
            allowed: this.metricAllowedRequests(),
            blocked: this.metricBlockedRequests(),
        }, "Blocked (rate)", undefined, undefined, this.region, this.account);
    }
}
exports.WafV2MetricFactory = WafV2MetricFactory;
_a = JSII_RTTI_SYMBOL_1;
WafV2MetricFactory[_a] = { fqn: "cdk-monitoring-constructs.WafV2MetricFactory", version: "8.1.0" };
