// cdk-monitoring-constructs/lib/monitoring/aws-sqs/SqsQueueMetricFactory.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.SqsQueueMetricFactory = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const common_1 = require("../../common");
class SqsQueueMetricFactory extends common_1.BaseMetricFactory {
    constructor(metricFactory, props) {
        super(metricFactory, props);
        this.queue = props.queue;
    }
    metricApproximateVisibleMessageCount() {
        return this.metricFactory.adaptMetric(this.queue.metricApproximateNumberOfMessagesVisible({
            label: "Visible",
            region: this.region,
            account: this.account,
        }));
    }
    metricIncomingMessageCount() {
        return this.metricFactory.adaptMetric(this.queue.metricNumberOfMessagesSent({
            statistic: common_1.MetricStatistic.SUM,
            label: "Incoming",
            region: this.region,
            account: this.account,
        }));
    }
    metricDeletedMessageCount() {
        return this.metricFactory.adaptMetric(this.queue.metricNumberOfMessagesDeleted({
            statistic: common_1.MetricStatistic.SUM,
            label: "Deleted",
            region: this.region,
            account: this.account,
        }));
    }
    metricApproximateAgeOfOldestMessageInSeconds() {
        return this.metricFactory.adaptMetric(this.queue.metricApproximateAgeOfOldestMessage({
            label: "Age",
            region: this.region,
            account: this.account,
        }));
    }
    metricAverageMessageSizeInBytes() {
        return this.metricFactory.adaptMetric(this.queue.metricSentMessageSize({
            label: "Size",
            region: this.region,
            account: this.account,
        }));
    }
    metricProductionRate() {
        return this.metricFactory.createMetricMath("(productionVolume / PERIOD(productionVolume))", {
            productionVolume: this.metricIncomingMessageCount(),
        }, "Production TPS (avg: ${AVG}, max: ${MAX})", undefined, undefined, this.region, this.account);
    }
    metricConsumptionRate() {
        return this.metricFactory.createMetricMath("(consumptionVolume / PERIOD(consumptionVolume))", {
            consumptionVolume: this.metricDeletedMessageCount(),
        }, "Consumption TPS (avg: ${AVG}, max: ${MAX})", undefined, undefined, this.region, this.account);
    }
    // Time to drain queue (number of visible messages / net consumption rate)
    // Net consumption rate is defined by consumption rate - incoming rate
    metricTimeToDrain() {
        return this.metricFactory.createMetricMath("(visibleMessages / (consumptionVolume - incomingVolume)) * (PERIOD(consumptionVolume))", {
            visibleMessages: this.metricApproximateVisibleMessageCount(),
            incomingVolume: this.metricIncomingMessageCount(),
            consumptionVolume: this.metricDeletedMessageCount(),
        }, "Time to Drain (seconds) (avg: ${AVG}, max: ${MAX})", undefined, undefined, this.region, this.account);
    }
}
exports.SqsQueueMetricFactory = SqsQueueMetricFactory;
_a = JSII_RTTI_SYMBOL_1;
SqsQueueMetricFactory[_a] = { fqn: "cdk-monitoring-constructs.SqsQueueMetricFactory", version: "8.1.0" };
