// cdk-monitoring-constructs/lib/monitoring/aws-loadbalancing/LoadBalancerMetricFactory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createLoadBalancerMetricFactory = void 0;
const ApplicationLoadBalancerMetricFactory_1 = require("./ApplicationLoadBalancerMetricFactory");
const NetworkLoadBalancerMetricFactory_1 = require("./NetworkLoadBalancerMetricFactory");
// It's not possible to use instanceOf with typescript interfaces, so we have to use type guards to differentiate
// between the interfaces. As another problem, the LoadBalancer/TargetGroup for both Network and Application types
// don't have distinguished fields that could be used to differentiate between both types, so we resort to using
// checking for unique methods in their metrics interfaces.
function isApplicationLoadBalancer(loadBalancer) {
    return !!loadBalancer.metrics.httpRedirectCount;
}
function isNetworkLoadBalancer(loadBalancer) {
    return !isApplicationLoadBalancer(loadBalancer);
}
function isApplicationTargetGroup(targetGroup) {
    return !!targetGroup.metrics.httpCodeTarget;
}
function isNetworkTargetGroup(targetGroup) {
    return !isApplicationTargetGroup(targetGroup);
}
/**
 * Factory method to create appropriate metric factory based on the load balancer and target group type.
 * @param metricFactory metric factory
 * @param loadBalancer load balancer
 * @param targetGroup target group
 */
function createLoadBalancerMetricFactory(metricFactory, loadBalancer, targetGroup, invertStatisticsOfTaskCountEnabled) {
    if (isNetworkLoadBalancer(loadBalancer) &&
        isNetworkTargetGroup(targetGroup)) {
        return new NetworkLoadBalancerMetricFactory_1.NetworkLoadBalancerMetricFactory(metricFactory, {
            networkLoadBalancer: loadBalancer,
            networkTargetGroup: targetGroup,
            invertStatisticsOfTaskCountEnabled: invertStatisticsOfTaskCountEnabled,
        });
    }
    else if (isApplicationLoadBalancer(loadBalancer) &&
        isApplicationTargetGroup(targetGroup)) {
        return new ApplicationLoadBalancerMetricFactory_1.ApplicationLoadBalancerMetricFactory(metricFactory, {
            applicationLoadBalancer: loadBalancer,
            applicationTargetGroup: targetGroup,
            invertStatisticsOfTaskCountEnabled: invertStatisticsOfTaskCountEnabled,
        });
    }
    else {
        throw new Error("Invalid type of load balancer or target group (only ALB and NLB are supported).");
    }
}
exports.createLoadBalancerMetricFactory = createLoadBalancerMetricFactory;
