// cdk-monitoring-constructs/lib/monitoring/aws-acm/CertificateManagerMonitoring.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CertificateManagerMonitoring = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const aws_cloudwatch_1 = require("aws-cdk-lib/aws-cloudwatch");
const CertificateManagerMetricFactory_1 = require("./CertificateManagerMetricFactory");
const common_1 = require("../../common");
const dashboard_1 = require("../../dashboard");
class CertificateManagerMonitoring extends common_1.Monitoring {
    constructor(scope, props) {
        super(scope, props);
        const namingStrategy = new dashboard_1.MonitoringNamingStrategy({ ...props });
        this.title = namingStrategy.resolveHumanReadableName();
        const metricFactory = new CertificateManagerMetricFactory_1.CertificateManagerMetricFactory(scope.createMetricFactory(), props);
        const alarmFactory = this.createAlarmFactory(namingStrategy.resolveAlarmFriendlyName());
        const ageAlarmFactory = new common_1.AgeAlarmFactory(alarmFactory);
        this.daysToExpiryAnnotations = [];
        this.daysToExpiryMetric = metricFactory.metricDaysToExpiry();
        for (const disambiguator in props.addDaysToExpiryAlarm) {
            const alarmProps = props.addDaysToExpiryAlarm[disambiguator];
            const createdAlarm = ageAlarmFactory.addDaysToExpiryAlarm(this.daysToExpiryMetric, alarmProps, disambiguator);
            this.daysToExpiryAnnotations.push(createdAlarm.annotation);
            this.addAlarm(createdAlarm);
        }
        props.useCreatedAlarms?.consume(this.createdAlarms());
    }
    summaryWidgets() {
        return [];
    }
    widgets() {
        return [
            this.createTitleWidget(),
            this.createDaysToExpiryWidget(common_1.QuarterWidth, common_1.DefaultGraphWidgetHeight),
        ];
    }
    createTitleWidget() {
        return new dashboard_1.MonitoringHeaderWidget({
            family: "Certificate",
            title: this.title,
        });
    }
    createDaysToExpiryWidget(width, height) {
        return new aws_cloudwatch_1.GraphWidget({
            width,
            height,
            title: "Expiration",
            left: [this.daysToExpiryMetric],
            leftAnnotations: this.daysToExpiryAnnotations,
            leftYAxis: common_1.CountAxisFromZero,
        });
    }
}
exports.CertificateManagerMonitoring = CertificateManagerMonitoring;
_a = JSII_RTTI_SYMBOL_1;
CertificateManagerMonitoring[_a] = { fqn: "cdk-monitoring-constructs.CertificateManagerMonitoring", version: "8.1.0" };
